pkg/cloudprovider/vsphereparavirtual/route.go

```diff
@@ -0,0 +1,415 @@
+	"k8s.io/cloud-provider-vsphere/pkg/util"
+	cloudnodeutil "k8s.io/cloud-provider/node/helpers"
+	klog "k8s.io/klog/v2"
+)
+
+// RoutesProvider is the interface definition for Routes functionality
+type RoutesProvider interface {
+	cloudprovider.Routes
+	AddNode(*v1.Node)
+	DeleteNode(*v1.Node)
+}
+
+type routesProvider struct {
+	routeClient client.Interface
+	namespace   string
+	nodeMap     map[string]*v1.Node
+	nodeMapLock sync.RWMutex
+	ownerRefs   []metav1.OwnerReference
+	kubeClient  clientset.Interface
+	nodeLister  listerv1.NodeLister
+}
+
+var _ RoutesProvider = &routesProvider{}
+
+const (
+	// AnnotationClusterName label is used to annotate GC name for Routes CR
+	AnnotationClusterName = v1alpha1.GroupName + "/clusterName"
+	// RealizedStateTimeout is the timeout duration for realized state check
+	// RealizedStateTimeout = 10 * time.Second
+	// RealizedStateSleepTime is the interval between realized state check
+	// RealizedStateSleepTime = 1 * time.Second
+)
+
+var updateNetworkConditionBackoff = wait.Backoff{
+	Steps:    5,
+	Duration: 100 * time.Millisecond,
+	Jitter:   1.0,
+}
+
+// A list of possible RouteSet operation error messages
+var (
+	ErrGetRouteSet    = errors.New("failed to get RouteSet")
+	ErrCreateRouteSet = errors.New("failed to create RouteSet")
+	ErrListRouteSet   = errors.New("failed to list RouteSet")
+	ErrDeleteRouteSet = errors.New("failed to delete RouteSet")
+)
+
+// GetRouteSetClient returns a new RouteSet client that can be used to access SC
+func GetRouteSetClient(config *rest.Config) (client.Interface, error) {
+	v1alpha1.AddToScheme(scheme.Scheme)
+	rClient, err := client.NewForConfig(config)
+	if err != nil {
+		klog.V(6).Infof("Failed to create RouteSet clientset")
+		return nil, err
+	}
+	return rClient, nil
+}
+
+// NewRoutes returns an implementation of RoutesProvider
+func NewRoutes(clusterNS string, kcfg *rest.Config, ownerRef metav1.OwnerReference, kubeClient clientset.Interface, nodeLister listerv1.NodeLister) (RoutesProvider, error) {
+	routeClient, err := GetRouteSetClient(kcfg)
+	if err != nil {
+		return nil, err
+	}
+
+	ownerRefs := []metav1.OwnerReference{
+		ownerRef,
+	}
+
+	return &routesProvider{
+		routeClient: routeClient,
+		namespace:   clusterNS,
+		nodeMap:     make(map[string]*v1.Node),
+		ownerRefs:   ownerRefs,
+		kubeClient:  kubeClient,
+		nodeLister:  nodeLister,
+	}, nil
+}
+
+// ListRoutes implements Routes.ListRoutes
+// Get RouteSet CR from SC namespace and then filters routes that belong to the specified clusterName
+// Only return cloudprovider.Route if RouteSet CR status 'Ready' is true
+func (r *routesProvider) ListRoutes(ctx context.Context, clusterName string) ([]*cloudprovider.Route, error) {
+	klog.V(6).Infof("Attempting to list Routes for cluster %s", clusterName)
+
+	routeSets, err := r.routeClient.NsxV1alpha1().RouteSets(r.namespace).List(ctx, metav1.ListOptions{})
```

Comment:
If adding clusterName in label, then you can add a labelSelector in `ListOptions`, then no need to filter RouteSet CRs which are belong to this cluster.

Reply:
yeah that's another option, I think either way works
if we use the labelSelector, is that okay for us to remove `clusterName` from annotations, does backend need this for other uses?

Reply:
It's fine, right now backend does not read RouteSet annotations.